Detectors/MUON/MCH/DevIO/Digits/digits-random-generator-workflow.cxx

```diff
@@ -0,0 +1,212 @@
+      throw std::invalid_argument("occupancy must be between >0 and <=1");
+    }
+    mNofRofPerTimeFrame = ic.options().get<int>(OPTNAME_NOF_ROFS_PER_TF);
+    mSeed = ic.options().get<int>(OPTNAME_SEED);
+
+    if (!mSeed) {
+      std::random_device rd;
+      mSeed = rd();
+    }
+    mMersenneTwister.seed(mSeed);
+
+    LOGP(INFO,
+         "Will generate {:7.2f}% of pads in {:4d} ROFs "
+         "per timeframe, for {:4d} timeframes",
+         mOccupancy * 100.0, mNofRofPerTimeFrame, mMaxNofTimeFrames);
+  }
+
+  void run(ProcessingContext& pc)
+  {
+    if (mNofProcessedTFs >= mMaxNofTimeFrames) {
+      pc.services().get<ControlService>().endOfStream();
+      return;
+    }
+
+    if (shouldProcess()) {
+      incNofProcessedTFs();
+      std::vector<Digit> digits;
+      std::vector<ROFRecord> rofs;
+      int ndigits{0};
+
+      for (auto i = 0; i < mNofRofPerTimeFrame; i++) {
+        auto n = generateRandomDigits(mOccupancy, digits);
+        o2::InteractionRecord ir{mBC, mOrbit++};
+        rofs.emplace_back(ir, ndigits, n);
+        ndigits += n;
+        mOrbit++;
+      }
+      pc.outputs().snapshot(OutputRef{"rofs"}, rofs);
+      pc.outputs().snapshot(OutputRef{"digits"}, digits);
+      printSummary(digits, rofs);
+      printFull(digits, rofs);
+    }
+    incTFid();
+  }
+
+  /**
+ * Populate the digits vector with (Nmch * occupancy) digits
+ * where Nmch is the total number of channels in MCH (1064008)
+ * 
+ * Each member variable of the digit is uniformly distributed within
+ * its expected range, as an attempt to maximize the entropy of the
+ * generated digits, which can be handy to test the efficiency of the
+ * entropy encoder used to create the Compressed Time Frame (CTF) for instance.
+ *
+ * @param occupancy is a number between 0 and 1.
+ * @digits a vector where the generated digits will be appended. 
+ * That vector is not cleared by this function, so digits can 
+ * be accumulated if need be.
+ * @returns the number of digits added to the input digits vector.
+ */
+  int generateRandomDigits(float occupancy,
+                           std::vector<Digit>& digits)
+  {
+    int n{0};
+    std::uniform_int_distribution<int32_t> adc{0, 1024 * 1024};
+    std::uniform_int_distribution<int32_t> tfTime{0, 512 * 3564};
+    std::uniform_int_distribution<int32_t> nofSamples{0, 1023};
+    std::uniform_real_distribution<float> sat{0.0, 1.0};
+
+    auto deids = getDeIds();
+    auto nofPadsPerDe = getNofPads();
+    auto& mt = mMersenneTwister;
+
+    for (auto i = 0; i < deids.size(); i++) {
+      auto deid = deids[i];
+      int16_t nofPads = static_cast<int16_t>(nofPadsPerDe[i]);
+      std::uniform_int_distribution<int16_t> padid{0, nofPads};
+      int nch = nofPads * occupancy;
+      for (int i = 0; i < nch; i++) {
+        auto p = padid(mt);
+        digits.emplace_back(deid, p, tfTime(mt), adc(mt), nofSamples(mt));
+        if (sat(mt) > 0.9) {
+          digits.back().setSaturated(true);
+        }
```

Comment:
this is a detail but you could pass (sat(mt) > 0.9) as the last argument of the emplace_back instead of changing the saturation bit afterward.

Reply:
indeed, when this code was first written, the Digit did not have the saturation in its constructor :wink: